PPSD: change internal representation of timestamps to integer nanosecond

megies opened the pull request:
This changes all internally used timestamps in PPSD to POSIX timestamps as integer nanoseconds, eliminating all potential floating point inaccuracies when e.g. comparing time stamps etc.

The only drawback is that npz files written on newer obspy versions will result in exceptions when reading on older obspy versions (the other way around should work fine). But I think the positive effect outweighs this drawback.

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [x] ~~If the PR is making changes to documentation, docs pages can be built automatically.
      Just remove the space in the following string after the + sign: "+ DOCS"~~
- [x] ~~If any network modules should be tested for the PR, add them as a comma separated list
      (e.g. `clients.fdsn,clients.arclink`) after the colon in the following magic string: "+TESTS:"
      (you can also add "ALL" to just simply run all tests across all modules)~~
- [x] All tests still pass.
- [x] ~~Any new features or fixed regressions are be covered via new tests.~~ no new features
- [x] ~~Any new or changed features have are fully documented.~~ no new features
- [x] Significant changes have been added to `CHANGELOG.txt` .
- [x] ~~First time contributors have added your name to `CONTRIBUTORS.txt` .~~

megies commented:
Also see #2034 for examples of unexpected behavior associated with floating point issues.

megies commented:
Good idea..

megies commented:
@krischer, I've added raising an exception in this case and will also do a PR for maintenance branch that at least shows a warning, so that the 1.1.x line is also covered with this..

krischer commented on obspy/signal/spectral_estimation.py:
```diff
@@ -1231,6 +1233,16 @@ def load_npz(filename, metadata=None):
+            msg = ("Trying to read a PPSD npz with 'ppsd_version={}'. This "
+                   "file was written on a more recent ObsPy version and can "
+                   "not safely be read with this ObsPy version (current "
+                   "'ppsd_version' is {}).").format(
```
Maybe add "Please update your ObsPy installation."?

krischer commented on CHANGELOG.txt:
```diff
@@ -5,6 +5,10 @@ master:
      objects to shapefile (see #2012)
+ - obspy.signal:
+   * PPSD: for safety reasons, raise an ObsPyException if trying to read a PPSD
+     npz file that was written with a newer ObsPy version than is currently
```
That's not really correct. Best rephrase to "written with a newer internal file format" or something like it. I guess the format does not change that often.

megies replied:
Yeah, you're right, that's misleading..

megies commented:
Made those wording changes and force-pushed.

megies commented:
Rebased and removed the exception for newer npz, was delegated to #2051. Hope I didn't screw up the merge conflict resolution, always a bit tricky when indent levels change (due to the context manager used now)

hash before rebasing: 7d684ba0127146c2a1ae94cb3dbdb09ce93d3c45

krischer commented:
Some tests fail now.

Jollyfant commented:
> Some tests fail now.

One of these is the test included in the PPSD fix we merged some days ago.

megies commented:
> One of these is the test included in the PPSD fix we merged some days ago.

Ah OK.. I rebased but didn't check that new test.. and somehow must've messed up the conflict resolution during rebasing, that npz version check somehow got lost.. -_-

Jollyfant commented:
It could be because of the change in `._times_processed` to nanoseconds representation.

megies commented:
OK, looks like this one's ready now. I'm a bit anxious about the internal time representation change.. but I think it's better in the long term and hopefully I thought about all issues that might arise when reading/adding older npz files..

krischer commented:
I fixed a conflict in the changelog and rebased. Can be merged from my point of view if CI is green.

megies commented:
Alright, let's do it.. 🙈
